src/http2_connection.erl

```diff
@@ -1016,26 +1018,46 @@ handle_sync_event(streams, _F, StateName,
     {reply, Streams, StateName, Conn};
 handle_sync_event({get_response, StreamId}, _F, StateName,
                   #connection{}=Conn) ->
-    Stream = get_stream(StreamId, Conn#connection.streams),
-    Reply = http2_stream:get_response(Stream#stream.pid),
-
+    Stream = h2_stream_set:get(StreamId, Conn#connection.streams),
+    Reply = case h2_stream_set:type(Stream) of
+                closed ->
+                    {ok, h2_stream_set:response(Stream)};
+                active ->
+                    not_ready
+            end,
     {reply, Reply, StateName, Conn};
 handle_sync_event({new_stream, NotifyPid}, _F, StateName,
                   #connection{
                      streams=Streams,
                      next_available_stream_id=NextId
                     }=Conn) ->
-
-    NewStream = new_stream_(NextId, NotifyPid, Conn),
-    lager:debug("[~p] added stream #~p to ~p",
-                [Conn#connection.type, NextId, Streams]),
-    {reply, NextId, StateName, Conn#connection{
+    {Reply, NewStreams} =
+        case
+            h2_stream_set:new_stream(
+              NextId,
+              NotifyPid,
+              Conn#connection.stream_callback_mod,
+              Conn#connection.socket,
+              Conn#connection.peer_settings#settings.initial_window_size,
+              Conn#connection.self_settings#settings.initial_window_size,
+              Streams)
+        of
+            {error, Code, _NewStream} ->
+                lager:warning("[~p] tried to create new_stream ~p, but there are too many",
```

Comment:
huh, interesting. what do we do if we want to make another request? spinlock until this returns ok?